File: sdks/official/fraiseql-csharp/src/FraiseQL/CrudGenerator.cs
using FraiseQL.Models;
using System.Text.RegularExpressions;

namespace FraiseQL;

/// <summary>
/// Generates standard CRUD queries and mutations for a GraphQL type.
/// </summary>
/// <remarks>
/// <para>Generated operations follow FraiseQL conventions:</para>
/// <list type="bullet">
/// <item>Read: query <c>&lt;snake&gt;</c> (get by PK) + query <c>&lt;snake&gt;s</c> (list with auto_params)</item>
/// <item>Create: mutation <c>create_&lt;snake&gt;</c> with a <c>Create{Type}Input</c> input object</item>
/// <item>Update: mutation <c>update_&lt;snake&gt;</c> with an <c>Update{Type}Input</c> input object</item>
/// <item>Delete: mutation <c>delete_&lt;snake&gt;</c> with PK only</item>
/// </list>
/// </remarks>
public static class CrudGenerator
{
    private static readonly Regex CamelRe = new(@"(?<!^)([A-Z])", RegexOptions.Compiled);

    /// <summary>Converts a PascalCase name to snake_case.</summary>
    /// <param name="name">The PascalCase name.</param>
    /// <returns>The snake_case equivalent.</returns>
    public static string PascalToSnake(string name) =>
        CamelRe.Replace(name, "_$1").ToLowerInvariant();

    /// <summary>
    /// Applies basic English pluralization rules to a snake_case name.
    /// </summary>
    /// <param name="name">The singular name.</param>
    /// <returns>The pluralized name.</returns>
    public static string Pluralize(string name)
    {
        if (name.EndsWith("s") && !name.EndsWith("ss")) return name;
        foreach (var suffix in new[] { "ss", "sh", "ch", "x", "z" })
            if (name.EndsWith(suffix)) return name + "es";
        if (name.Length >= 2 && name[^1] == 'y' && !"aeiou".Contains(name[^2]))
            return name[..^1] + "ies";
        return name + "s";
    }

    /// <summary>
    /// Generates CRUD queries and mutations for the given type.
    /// </summary>
    /// <param name="typeName">The GraphQL type name (PascalCase).</param>
    /// <param name="fields">The fields on this type. The first field is assumed to be the primary key.</param>
    /// <param name="sqlSource">Optional SQL source override. Defaults to <c>v_&lt;snake&gt;</c>.</param>
    /// <param name="cascade">When <see langword="true"/>, generated mutations include cascade support.</param>
    /// <returns>A tuple of generated queries, mutations, and input types.</returns>
    /// <exception cref="InvalidOperationException">Thrown when <paramref name="fields"/> is empty.</exception>
    public static (IReadOnlyList<IntermediateQuery> Queries, IReadOnlyList<IntermediateMutation> Mutations, IReadOnlyList<InputTypeDefinition> InputTypes)
        Generate(string typeName, IReadOnlyList<IntermediateField> fields, string? sqlSource = null, bool cascade = false)
    {
        if (fields.Count == 0)
            throw new InvalidOperationException(
                $"Type '{typeName}' has no fields; cannot generate CRUD operations");

        var snake = PascalToSnake(typeName);
        var view = string.IsNullOrEmpty(sqlSource) ? $"v_{snake}" : sqlSource;
        var pkField = fields[0];
        bool? cascadeValue = cascade ? true : null;

        var queries = new List<IntermediateQuery>
        {
            // Get-by-ID query
            new(
                Name: snake,
                ReturnType: typeName,
                ReturnsList: false,
                Nullable: true,
                SqlSource: view,
                Arguments: new[] { new IntermediateArgument(pkField.Name, pkField.Type, false) }.ToList().AsReadOnly(),
                Description: $"Get {typeName} by ID."),

            // List query
            new(
                Name: Pluralize(snake),
                ReturnType: typeName,
                ReturnsList: true,
                Nullable: false,
                SqlSource: view,
                Arguments: Array.Empty<IntermediateArgument>().ToList().AsReadOnly(),
                Description: $"List {typeName} records.")
        };

        var mutations = new List<IntermediateMutation>();
        var inputTypes = new List<InputTypeDefinition>();

        // Create mutation — input object with all non-computed fields
        var createInputName = $"Create{typeName}Input";
        var createInputFields = fields
            .Where(f => f.Computed != true)
            .Select(f => new InputFieldDefinition(f.Name, f.Type, f.Nullable))
            .ToList()
            .AsReadOnly();
        inputTypes.Add(new InputTypeDefinition(
            createInputName, createInputFields, $"Input for creating a new {typeName}."));

        mutations.Add(new IntermediateMutation(
            Name: $"create_{snake}",
            ReturnType: typeName,
            SqlSource: $"fn_create_{snake}",
            Operation: "INSERT",
            Arguments: new[] { new IntermediateArgument("input", createInputName, false) }.ToList().AsReadOnly(),
            Description: $"Create a new {typeName}.",
            Cascade: cascadeValue));

        // Update mutation — input object with PK required, rest nullable
        var updateInputName = $"Update{typeName}Input";
        var updateInputFields = new List<InputFieldDefinition>
        {
            new(pkField.Name, pkField.Type, false)
        };
        updateInputFields.AddRange(fields.Skip(1).Where(f => f.Computed != true).Select(f => new InputFieldDefinition(f.Name, f.Type, true)));
        inputTypes.Add(new InputTypeDefinition(
            updateInputName, updateInputFields.AsReadOnly(), $"Input for updating an existing {typeName}."));

        mutations.Add(new IntermediateMutation(
            Name: $"update_{snake}",
            ReturnType: typeName,
            SqlSource: $"fn_update_{snake}",
            Operation: "UPDATE",
            Arguments: new[] { new IntermediateArgument("input", updateInputName, false) }.ToList().AsReadOnly(),
            Description: $"Update an existing {typeName}.",
            Cascade: cascadeValue));

        // Delete mutation — PK only (no input object)
        mutations.Add(new IntermediateMutation(
            Name: $"delete_{snake}",
            ReturnType: typeName,
            SqlSource: $"fn_delete_{snake}",
            Operation: "DELETE",
            Arguments: new[] { new IntermediateArgument(pkField.Name, pkField.Type, false) }.ToList().AsReadOnly(),
            Description: $"Delete a {typeName}.",
            Cascade: cascadeValue));

        return (queries.AsReadOnly(), mutations.AsReadOnly(), inputTypes.AsReadOnly());
    }

    /// <summary>
    /// Represents a field within a generated input type.
    /// </summary>
    /// <param name="Name">The field name.</param>
    /// <param name="Type">The GraphQL type name.</param>
    /// <param name="Nullable">Whether this field accepts null.</param>
    public record InputFieldDefinition(string Name, string Type, bool Nullable);

    /// <summary>
    /// Represents a generated input type (e.g. <c>CreateUserInput</c>, <c>UpdateUserInput</c>).
    /// </summary>
    /// <param name="Name">The input type name.</param>
    /// <param name="Fields">Ordered list of fields on this input type.</param>
    /// <param name="Description">Human-readable description.</param>
    public record InputTypeDefinition(
        string Name,
        IReadOnlyList<InputFieldDefinition> Fields,
        string Description);
}

File: sdks/official/fraiseql-fsharp/src/FraiseQL/CrudGenerator.fs
namespace FraiseQL

open System.Text.RegularExpressions

/// Generates standard CRUD queries and mutations for a GraphQL type.
///
/// Generated operations follow FraiseQL conventions:
///   - Read:   query <snake> (get by PK) + query <snake>s (list with auto_params)
///   - Create: mutation create_<snake> with a Create<Type>Input input object
///   - Update: mutation update_<snake> with an Update<Type>Input input object
///   - Delete: mutation delete_<snake> with PK only
module CrudGenerator =

    let private camelRe = Regex(@"(?<!^)([A-Z])", RegexOptions.Compiled)

    /// Converts a PascalCase name to snake_case.
    let pascalToSnake (name: string) =
        camelRe.Replace(name, "_$1").ToLowerInvariant()

    /// Applies basic English pluralization rules to a snake_case name.
    let pluralize (name: string) =
        if name.EndsWith("s") && not (name.EndsWith("ss")) then
            name
        elif
            [ "ss"; "sh"; "ch"; "x"; "z" ]
            |> List.exists name.EndsWith
        then
            name + "es"
        elif
            name.Length >= 2
            && name.[name.Length - 1] = 'y'
            && not ("aeiou".Contains(name.[name.Length - 2]))
        then
            name.[.. name.Length - 2] + "ies"
        else
            name + "s"

    /// Generates CRUD queries, mutations, and input types for the given type.
    ///
    /// Returns a tuple of (queries, mutations, input_types).
    ///
    /// # Errors
    ///
    /// Raises System.InvalidOperationException when the fields list is empty.
    let generate
        (typeName: string)
        (fields: FieldDefinition list)
        (sqlSource: string)
        (cascade: bool)
        : QueryDefinition list * MutationDefinition list * InputTypeDefinition list
        =
        if fields.IsEmpty then
            raise (
                System.InvalidOperationException(
                    sprintf "Type '%s' has no fields; cannot generate CRUD operations" typeName
                )
            )

        let snake = pascalToSnake typeName

        let view =
            if System.String.IsNullOrEmpty(sqlSource) then
                "v_" + snake
            else
                sqlSource

        let pkField = fields.[0]
        let cascadeValue = if cascade then Some true else None

        let createInputName = sprintf "Create%sInput" typeName

        let createInputType: InputTypeDefinition =
            {
                name = createInputName
                fields =
                    fields
                    |> List.filter (fun f -> not f.computed)
                    |> List.map (fun f ->
                        {
                            name = f.name
                            type_ = f.type_
                            nullable = f.nullable
                        })
                description = Some(sprintf "Input for creating a new %s." typeName)
            }

        let updateInputName = sprintf "Update%sInput" typeName

        let updateInputType: InputTypeDefinition =
            {
                name = updateInputName
                fields =
                    { name = pkField.name; type_ = pkField.type_; nullable = false }
                    :: (fields
                        |> List.tail
                        |> List.filter (fun f -> not f.computed)
                        |> List.map (fun f ->
                            {
                                name = f.name
                                type_ = f.type_
                                nullable = true
                            }))
                description = Some(sprintf "Input for updating an existing %s." typeName)
            }

        let queries =
            [
                // Get-by-ID query
                {
                    name = snake
                    return_type = typeName
                    returns_list = false
                    nullable = true
                    sql_source = view
                    sql_source_dispatch = None
                    arguments =
                        [
                            {
                                name = pkField.name
                                type_ = pkField.type_
                                nullable = false
                            }
                        ]
                    cache_ttl_seconds = None
                    description = Some(sprintf "Get %s by ID." typeName)
                    rest = None
                }
                // List query
                {
                    name = pluralize snake
                    return_type = typeName
                    returns_list = true
                    nullable = false
                    sql_source = view
                    sql_source_dispatch = None
                    arguments = []
                    cache_ttl_seconds = None
                    description = Some(sprintf "List %s records." typeName)
                    rest = None
                }
            ]

        let mutations =
            [
                // Create mutation — single input object argument
                {
                    name = "create_" + snake
                    return_type = typeName
                    sql_source = "fn_create_" + snake
                    sql_source_dispatch = None
                    operation = "INSERT"
                    arguments =
                        [
                            {
                                name = "input"
                                type_ = createInputName
                                nullable = false
                            }
                        ]
                    description = Some(sprintf "Create a new %s." typeName)
                    rest = None
                    cascade = cascadeValue
                }
                // Update mutation — single input object argument
                {
                    name = "update_" + snake
                    return_type = typeName
                    sql_source = "fn_update_" + snake
                    sql_source_dispatch = None
                    operation = "UPDATE"
                    arguments =
                        [
                            {
                                name = "input"
                                type_ = updateInputName
                                nullable = false
                            }
                        ]
                    description = Some(sprintf "Update an existing %s." typeName)
                    rest = None
                    cascade = cascadeValue
                }
                // Delete mutation — PK only (unchanged)
                {
                    name = "delete_" + snake
                    return_type = typeName
                    sql_source = "fn_delete_" + snake
                    sql_source_dispatch = None
                    operation = "DELETE"
                    arguments =
                        [
                            {
                                name = pkField.name
                                type_ = pkField.type_
                                nullable = false
                            }
                        ]
                    description = Some(sprintf "Delete a %s." typeName)
                    rest = None
                    cascade = cascadeValue
                }
            ]

        let inputTypes = [ createInputType; updateInputType ]

        (queries, mutations, inputTypes)
